src/xhydro/frequency_analysis/uncertainities.py

```diff
@@ -0,0 +1,255 @@
+
+    Returns
+    -------
+    xarray.Dataset
+        L-moments for each bootstrap sample.
+    """
+    ds_mom = []
+    for sample in ds_samples.samples.values:
+        ds = ds_samples.sel(samples=sample)
+        ds_mom.append(
+            xr.apply_ufunc(
+                moment_l_vector,
+                ds,
+                input_core_dims=[["time"]],
+                output_core_dims=[["lmom"]],
+            ).assign_coords(lmom=["l1", "l2", "l3", "tau", "tau3", "tau4"])
+        )
+    return xr.concat(ds_mom, dim="samples")
+
+
+def calc_q_iter(
+    bv, var, ds_groups, ds_moments_iter, return_periods, small_regions_threshold=5
+):
+    """
+    Calculate quantiles for each bootstrap sample and group.
+
+    Parameters
+    ----------
+    bv : str
+        The basin identifier.
+    var : str
+        The variable name.
+    ds_groups : xarray.Dataset
+        The grouped data.
+    ds_moments_iter : xarray.Dataset
+        The L-moments for each bootstrap sample.
+    return_periods : array-like
+        The return periods to calculate quantiles for.
+    small_regions_threshold : int, optional
+        The threshold for removing small regions. Default is 5.
+
+    Returns
+    -------
+    xarray.DataArray
+        Quantiles for each bootstrap sample and group.
+    """
+    # We select groups for only one id
+    ds_temp = ds_groups[[var]].sel(id=bv).dropna("group_id", "all")
+    ds_mom = []
+
+    # For each group, we find which id are in it
+    for group_id in ds_temp.group_id.values:
+        id_list = ds_groups.sel(group_id=group_id).dropna("id", how="all").id.values
+        # We use moments with ressample previously done, and we create ds_moment_group with iterations
+        ds_mom.append(
+            ds_moments_iter[[var]]
+            .sel(id=id_list)
+            .assign_coords(group_id=group_id)
+            .expand_dims("group_id")
+        )
+
+    # Concat along group_id
+    ds_moments_groups = xr.concat(ds_mom, dim="group_id")
+    ds_groups = (
+        ds_groups[[var]]
+        .sel(group_id=ds_moments_groups.group_id.values)
+        .dropna(dim="id", how="all")
+    )
+    # With obs and moments  of same dims, we calculate
+    qt = calculate_rp_from_afr(ds_groups, ds_moments_groups, return_periods)
+    qt = remove_small_regions(qt, thresh=small_regions_threshold)
+    # pour chaque bv on stack les régions et bootstat et on calcul les centiles
+    return (
+        qt.rename({"samples": "obs_samples"})
+        .stack(samples=["group_id", "obs_samples"])
+        .squeeze()
+    )
+
+
+def generate_combinations(da, n):
+    """
+    Generate combinations of indices omitting up to N indices.
+
+    Parameters
+    ----------
+    da : (xarray.DataArray):
+        Input DataArray.
+    n : (int)
+        Number of indices to omit in each combination.
+
+    Returns
+    -------
+    list
+        List of all combinations.
+    """
+    # Get the list of indices
+    for id in da.indexes:
```

Comment:
```suggestion
    for ids in da.indexes:
```
Ou peu importe ce que tu préfères. `id` est un terme de base dans python, donc idéalement on n'y touche pas: https://www.w3schools.com/python/ref_func_id.asp